Implement referential integrity validation for Search Relevance entities

Bharathi-Kanna opened the pull request:
### Description
This PR implements referential integrity validation for the Search Relevance plugin. It ensures that whenever a user creates or updates an entity, all referenced resources (indices, query sets, search configurations, and judgments) are verified for existence before the operation proceeds.


### Issues Resolved
Closes #359
<img width="1440" height="815" alt="Screenshot 2025-12-25 at 11 54 04 PM" src="https://github.com/user-attachments/assets/0c71af1b-e55c-4cf5-8e39-ec19d0739e34" />
<img width="1440" height="810" alt="Screenshot 2025-12-25 at 11 53 29 PM" src="https://github.com/user-attachments/assets/5ed73d05-b9a1-4634-8e2f-693c1e1128e9" />


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

iprithv reviewed (commented):
LGTM 🚢

martin-gaievski commented on src/main/java/org/opensearch/searchrelevance/transport/experiment/PutExperimentTransportAction.java:
```diff
@@ -108,10 +120,52 @@ protected void doExecute(Task task, PutExperimentRequest request, ActionListener
+        );
+
+        // Validate Search Configurations
+        if (request.getSearchConfigurationList() != null) {
```
nit: and check for empty list, this way it's consistent with other validation in this method

martin-gaievski commented on CHANGELOG.md:
```diff
@@ -8,7 +8,8 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - adds version-based index mapping update support to the Search Relevance plugin [#344](https://github.com/opensearch-project/search-relevance/pull/344)
-- LLM Judgement Customized Prompt Template Implementation  [#264](https://github.com/opensearch-project/search-relevance/pull/264)
+* LLM Judgement Customized Prompt Template Implementation  [#264](https://github.com/opensearch-project/search-relevance/pull/264)
+* Implement referential integrity validation for search configurations, experiments, and judgments ([#360](https://github.com/opensearch-project/search-relevance/pull/360))
```
this change looks more like a bug fix

martin-gaievski reviewed (commented):
Thanks for this contribution, please check my comments, they all are minor I believe. In addition to those comments, please consider adding negative integration tests per API (lightweight, no data ingestion needed, just `expectThrows(ResponseException.class)` with fake IDs. Such integ tests are still light and will increase confidence

Bharathi-Kanna committed:
Fix: Add referential integrity validation

Adds validation to ensure QuerySets, Search Configurations, and
Judgments referenced by new Experiments and Judgments exist.
Includes negative integration tests and resolves PR feedback.

Signed-off-by: Bharathi-Kanna <99189546+Bharathi-Kanna@users.noreply.github.com>

Bharathi-Kanna commented:
@iprithv @martin-gaievski Thanks for the review! 
@martin-gaievski, I have updated the PR respectively.

martin-gaievski commented on src/test/java/org/opensearch/searchrelevance/action/ReferenceValidationIT.java:
```diff
@@ -0,0 +1,180 @@
+            + "\"size\": 10"
+            + "}";
+
+        expectThrows(
```
here you're sending `"searchConfigurationList": []` `RestPutExperimentAction` validates that `searchConfigurationList` cannot be empty and throws `IllegalArgumentException` before reaching the transport layer. test will fail with the wrong exception, it'll get an `IllegalArgumentException` about empty searchConfigurationList, not a `ResponseException` about non-existent querySetId.

Same comment for method `testExperimentWithFakeJudgmentId_thenFails`

martin-gaievski commented on src/main/java/org/opensearch/searchrelevance/utils/ReferenceValidationUtil.java:
```diff
@@ -0,0 +1,95 @@
+     * @param indexName      Index name to validate
+     * @return ValidationResult indicating if the index exists
+     */
+    public static ValidationResult validateIndexExists(ClusterService clusterService, String indexName) {
```
nit: seems this method is only used in tests, consider moving it from production code to some test util class

martin-gaievski commented on src/test/java/org/opensearch/searchrelevance/action/judgment/LlmJudgmentTemplateIT.java:
```diff
@@ -337,6 +355,16 @@ public void testLlmJudgmentWithOverwriteCache_thenSuccessful() {
+        String bulkData = Files.readString(Path.of(classLoader.getResource("llmjudgment/BulkIngestProducts.json").toURI()));
+        bulkIngest(TEST_INDEX, bulkData);
+
+        // Step 3: Create query set
```
nit: duplicate of next string, please remove one the comments

Bharathi-Kanna committed:
Address maintainer feedback for referential validation

Signed-off-by: Bharathi-Kanna <99189546+Bharathi-Kanna@users.noreply.github.com>

Bharathi-Kanna committed:
Remove PR number reference from comment

Signed-off-by: Bharathi-Kanna <99189546+Bharathi-Kanna@users.noreply.github.com>

Bharathi-Kanna commented:
Hey @martin-gaievski , I've addressed all your review comments  
 
I also noticed the recent CI run failed specifically on the `Rolling-Upgrade BWC Tests`. This is expected! Since our PR actively tightens referential integrity, old BWC tests that lazily try to pass dummy string IDs are now correctly getting blocked with our `400 Bad Request`. Do you want me to update the old BWC scripts to create valid configs first, or should we just adjust their expected assertions to handle the `400` response?
